client/src/actions/annotation.ts

```diff
@@ -0,0 +1,631 @@
+      categoryName,
+      prevObsCrossfilter.allSelectedLabels(),
+      labelName
+    );
+
+    dispatch({
+      type: "annotation: label current cell selection",
+      metadataField: categoryName,
+      label: labelName,
+      obsCrossfilter,
+      annoMatrix: obsCrossfilter.annoMatrix,
+    });
+  };
+
+function writableAnnotations(
+  annoMatrix: AnnoMatrix | undefined | null
+): string[] {
+  if (!annoMatrix) return [];
+  return (annoMatrix.schema?.annotations?.obs?.columns ?? [])
+    .filter((column) => column.writable)
+    .map((column) => column.name);
+}
+
+export function changedWritableAnnotations(
+  annoMatrix: AnnoMatrix | undefined | null,
+  lastSavedAnnoMatrix: AnnoMatrix | null
+): string[] {
+  if (!annoMatrix || !lastSavedAnnoMatrix) return [];
+
+  const currentBase = annoMatrix.base();
+  const lastBase = lastSavedAnnoMatrix.base();
+
+  if (currentBase === lastBase) return [];
+
+  const currentWritable = writableAnnotations(currentBase);
+  const lastWritable = writableAnnotations(lastBase);
+
+  const changed = new Set<string>();
+  // Columns newly added since last save
+  for (const col of currentWritable) {
+    if (!lastWritable.includes(col)) changed.add(col);
+  }
+  // Determine which existing columns have modified data
+  const currentObsCache = currentBase._cache?.[Field.obs];
+  const lastObsCache = lastBase._cache?.[Field.obs];
+
+  // If caches are identical, then nothing changed beyond new/removed columns
+  if (currentObsCache !== lastObsCache) {
+    for (const col of currentWritable) {
+      const currentHas = currentObsCache?.hasCol?.(col) ?? false;
+      const lastHas = lastObsCache?.hasCol?.(col) ?? false;
+      if (currentHas && lastHas) {
+        if (currentObsCache.col(col) !== lastObsCache.col(col))
+          changed.add(col);
+      } else if (currentHas && !lastHas) {
+        changed.add(col);
+      }
+    }
+  }
+
+  return Array.from(changed);
+}
+
+export function deletedWritableAnnotations(
+  annoMatrix: AnnoMatrix | undefined | null,
+  lastSavedAnnoMatrix: AnnoMatrix | null
+): string[] {
+  if (!annoMatrix || !lastSavedAnnoMatrix) return [];
+
+  const currentBase = annoMatrix.base();
+  const lastBase = lastSavedAnnoMatrix.base();
+
+  if (currentBase === lastBase) return [];
+
+  const currentWritable = writableAnnotations(currentBase);
+  const lastWritable = writableAnnotations(lastBase);
+
+  // Find columns that existed in last save but no longer exist
+  const deleted: string[] = [];
+  for (const col of lastWritable) {
+    if (!currentWritable.includes(col)) {
+      deleted.push(col);
+    }
+  }
+
+  return deleted;
+}
+
+export const needToSaveObsAnnotations = (
+  annoMatrix: AnnoMatrix | undefined,
+  lastSavedAnnoMatrix: AnnoMatrix | null
+): boolean =>
+  changedWritableAnnotations(annoMatrix, lastSavedAnnoMatrix).length > 0 ||
+  deletedWritableAnnotations(annoMatrix, lastSavedAnnoMatrix).length > 0;
+
+export const saveObsAnnotationsAction =
```

Comment:
Autosave orchestration lives here: we diff the current annoMatrix against the last saved snapshot, encode changed columns to FBS, gzip them (`pako.deflate`), and PUT to the new endpoint. If nothing changed except deletions we bail early to avoid uploading empty payloads.